mathoxide-lib/src/utils.rs

```diff
@@ -0,0 +1,77 @@
+pub struct ByRowIterator {
+    shape: Vec<usize>,
+    current: usize,
+    numel: usize,
+}
+
+impl Iterator for ByRowIterator {
+    type Item = Vec<usize>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.current >= self.numel {
+            None
+        } else {
+            let v = inverse_translate(self.current, &self.shape);
+            self.current += 1;
+            Some(v)
+        }
+    }
+}
+
+fn inverse_translate<Shape: AsRef<[usize]>>(idx: usize, shape: Shape) -> Vec<usize> {
+    shape.as_ref().iter().rev().scan((1, 0), |(divisor, remainder), e| {
+        let prev_div = *divisor;
+        *divisor *= e;
+        let res = ((idx - *remainder) % *divisor) / prev_div;
+        *remainder += res;
+        Some(res)
+    }).collect()
+}
+
+
+pub fn iterate_by_column<Shape: AsRef<[usize]>>(shape: &Shape) -> ByRowIterator {
```

Comment:
Maybe the name could be `iterate_row_major`. By column could be confused as going over the whole columns one by one in a column major order